### PR TITLE
Resolve testhost from source directory if we couldnt resolve it via nuget cache

### PR DESCRIPTION
**Pull request by Faizan2304**
Issue: https://github.com/Microsoft/vstest/issues/657 

Fix: Try to resolve testhost from source location if you couldn’t resolve it via nuget.

Test: Added UT to cover change and tested the fix manually

### CODE REVIEW

File: test/Microsoft.TestPlatform.TestHostProvider.UnitTests/Hosting/DotnetTestHostManagerTests.cs [hunk 1]
```diff
@@ -426,6 +426,73 @@ public void GetTestHostProcessStartInfoShouldIncludeTestHostPathFromDepsFile()
...
+            this.mockFileHelper.Setup(ph => ph.Exists("\\tmp\\test.runtimeconfig.dev.json")).Returns(true);
+
+            MemoryStream depsFileStream = new MemoryStream(Encoding.UTF8.GetBytes(depsFileContent));
+            this.mockFileHelper.Setup(ph => ph.GetStream("\\tmp\\test.deps.json", FileMode.Open, FileAccess.ReadWrite)).Returns(depsFileStream);
+            this.mockFileHelper.Setup(ph => ph.Exists("\\tmp\\test.deps.json")).Returns(true);
+
+            string testHostFullPath = @"C:\packages\microsoft.testplatform.testhost/15.0.0-Dev\lib/netstandard1.5/testhost.dll";
+            this.mockFileHelper.Setup(ph => ph.Exists(testHostFullPath)).Returns(false);
+
+            string testHostPath = Path.Combine($"{Path.DirectorySeparatorChar}tmp", "testhost.dll");
```
Comment (harshjain2) [inline test/Microsoft.TestPlatform.TestHostProvider.UnitTests/Hosting/DotnetTestHostManagerTests.cs]: >string testHostPath = Path.Combine($"{Path.DirectorySeparatorChar}tmp", "testhost.dll"); [](start = 12, length = 88)

Reuse class variable declared above: 
Line No : 63 : this.defaultTestHostPath = @"\tmp\testhost.dll";

**Comment by harshjain2** [comment]
:shipit:

**Review by harshjain2 — APPROVED** [review]
👍 with a suggestion.

File: src/Microsoft.TestPlatform.TestHostProvider/Hosting/DotnetTestHostManager.cs [hunk 2]
```diff
@@ -394,14 +394,12 @@ private string GetTestHostPath(string runtimeConfigDevPath, string depsFilePath,
...
-                // and is running tests in an isolated machine. A second scenario is self test: test platform unit tests take a project
-                // dependency on testhost (instead of nuget dependency), this drops testhost to output path.
-                testHostPath = Path.Combine(sourceDirectory, "testhost.dll");
-                EqtTrace.Verbose("DotnetTestHostManager: Assume published test project, with test host path = {0}.", testHostPath);
-            }
+            // If we are here it means it couldnt resolve testhost.dll from nuget chache.
+            // Try resolving testhost from output directory of test project. This is required if user has published the test project
+            // and is running tests in an isolated machine. A second scenario is self test: test platform unit tests take a project
+            // dependency on testhost (instead of nuget dependency), this drops testhost to output path.
+            testHostPath = Path.Combine(sourceDirectory, "testhost.dll");
```
Comment (smadala) [inline src/Microsoft.TestPlatform.TestHostProvider/Hosting/DotnetTestHostManager.cs]: We should handle case if testhost.dll does not exist in sourceDirectory too for some reason.

Comment (AbhitejJohn) [inline src/Microsoft.TestPlatform.TestHostProvider/Hosting/DotnetTestHostManager.cs]: Interesting. Would this publish the Extensions folder, that has the default extensions here too?

**Review by codito — APPROVED** [review]

File: test/Microsoft.TestPlatform.TestHostProvider.UnitTests/Hosting/DotnetTestHostManagerTests.cs [hunk 3]
```diff
@@ -426,6 +426,73 @@ public void GetTestHostProcessStartInfoShouldIncludeTestHostPathFromDepsFile()
...
+        ""type"": ""package"",
+        ""serviceable"": true,
+        ""sha512"": ""sha512-enO8sZmjbhXOfiZ6hV2ncaknaHnQbrGVsHUJzzu2Dmoh4fHFro4BF1Y4+sb4LOQhu4b3DFYPRj1ncd1RQK6HmQ=="",
+        ""path"": ""microsoft.testplatform.testhost/15.0.0-Dev"",
+        ""hashPath"": ""microsoft.testplatform.testhost.15.0.0-Dev""
+        }
+    }
+}";
+
+            MemoryStream runtimeConfigStream = new MemoryStream(Encoding.UTF8.GetBytes(runtimeConfigFileContent));
```
Comment (codito) [inline test/Microsoft.TestPlatform.TestHostProvider.UnitTests/Hosting/DotnetTestHostManagerTests.cs]: Please use `using...`

File: src/Microsoft.TestPlatform.TestHostProvider/Hosting/DotnetTestHostManager.cs (hunk 2 above)
Comment (Faizan2304) [inline src/Microsoft.TestPlatform.TestHostProvider/Hosting/DotnetTestHostManager.cs]: @smadala : We are handling the case if testhost.dll doesn't exist here https://github.com/Microsoft/vstest/blob/master/src/Microsoft.TestPlatform.TestHostProvider/Hosting/DotnetTestHostManager.cs#L243

Comment (Faizan2304) [inline src/Microsoft.TestPlatform.TestHostProvider/Hosting/DotnetTestHostManager.cs]: @AbhitejJohn : very interesting catch 👍 . we are planning to handle thing related to Extension in separate task.

File: test/Microsoft.TestPlatform.TestHostProvider.UnitTests/Hosting/DotnetTestHostManagerTests.cs (hunk 3 above)
Comment (Faizan2304) [inline test/Microsoft.TestPlatform.TestHostProvider.UnitTests/Hosting/DotnetTestHostManagerTests.cs]: We cant use using here as these memory stream are getting use in mock

**Pull request merged**